platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrClientAdapter.java

```diff
@@ -0,0 +1,998 @@
+      }
+      final Object status = response.getResponse().get("status");
+
+      if (SolrClientAdapter.OK_STATUS.equals(status)) {
+        setConnected(true);
+      } else {
+        LOGGER.debug(SolrClientAdapter.FAILED_TO_PING_WITH_STATUS, core, status);
+        setConnecting(
+            realClient,
+            new UnavailableSolrClient(
+                new UnavailableSolrException("ping failed with " + status + " status")),
+            true,
+            State.CONNECTED);
+      }
+      return response;
+    } catch (UnavailableSolrException | VirtualMachineError e) {
+      throw e;
+    } catch (Throwable t) {
+      LOGGER.debug(SolrClientAdapter.FAILED_TO_PING, core, t, t);
+      setConnecting(realClient, new UnavailableSolrClient(t), true, State.CONNECTED);
+      throw t;
+    }
+  }
+
+  private void checkIfReachable(
+      @SuppressWarnings({"unused", "squid:S1172"} /* required by failsafe's API */)
+          ExecutionContext context) {
+    checkIfReachable("from failsafe while trying to reconnect");
+  }
+
+  private void finalizeStateChange(
+      boolean notifyAvailability,
+      @Nullable Future<?> futureToCancel,
+      @Nullable SolrClient previousClientToClose,
+      boolean swallowIOExceptions)
+      throws IOException {
+    if (notifyAvailability) {
+      notifyListenersAndInitializers();
+    }
+    if ((futureToCancel != null) && !futureToCancel.isDone()) {
+      LOGGER.debug("Solr({}): cancelling its previous failsafe task", core);
+      futureToCancel.cancel(true);
+    }
+    // don't close if we still use the same client
+    if ((previousClientToClose != null) && (previousClientToClose != realClient)) {
+      LOGGER.debug("Solr({}): closing its previous client [{}]", core, previousClientToClose);
+      Closeables.close(previousClientToClose, swallowIOExceptions);
+    }
+  }
+
+  private void notifyListenersAndInitializers() {
+    final boolean available = (state == State.CONNECTED);
+    final String availableString = SolrClientAdapter.availableToString(available);
+
+    if (LOGGER.isInfoEnabled()) {
+      if (state == State.CLOSED) {
+        LOGGER.info("Solr client ({}) is closed", core);
+      } else {
+        LOGGER.info("Solr client ({}) is {}", core, availableString.toLowerCase());
+      }
+    }
+    if (!listeners.isEmpty()) {
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug(
+            "Solr({}): starting background task(s) to notify {} listener(s) that the availability changed to [{}]",
+            core,
+            listeners.size(),
+            availableString);
+      }
+      listeners.forEach(l -> executor.submit(() -> notifyAvailability(l, "changed to")));
+    }
+    if (available && !initializers.isEmpty()) {
+      Initializer i;
+
+      LOGGER.debug(
+          "Solr({}): starting background task(s) to notify {} initializer(s)",
+          core,
+          initializers.size());
+      while ((i = initializers.poll()) != null) {
```

Comment:
Availability is by definition something that can change at any time. There is nothing we can do about that. Therefore, even as we are notifying a listener that the client became available, it is entirely possible that by the time the call back into the client, it is no longer available. Same is true if you check the availability manually or even with an active ping, the client might not be available by the time the next call comes in. As for notifications, we might be sending multiple notifications for sure, I do not see any way out of that would keep the code simple enough. As a  result, we trigger the notification and leave it up to the notification thread to check the state before calling back the listener. 

Initializers are much simpler because they only indicate that we became available and not that we are. This is why, the code here removes each initializer as they are being notified.

So pretty much all state changes path leads to this method (actually to the `finalizeStateChange()` method which calls this one if it was detected that the availability somehow changed. That check is always done under synchronization.